tensorflow/core/kernels/roll_op.cc

```diff
@@ -292,47 +102,266 @@ class RollOp : public OpKernel {
     auto input_flat = input.flat<T>().data();
     auto output_flat = output->flat<T>().data();
 
-    if (std::is_same<Device, CPUDevice>::value) {
-      if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
-        // V2 copies memory in groups instead of element by element
-        DoRollWithMemcpy<T>(context, num_elements, num_dims, dim_size,
-                            input_flat, output_flat, threshold, dim_range, isd);
-      } else {
-        // incase memcpy does not work for current data type
-        DoRoll<T>(context, num_elements, num_dims, dim_size, input_flat,
-                  output_flat, threshold, dim_range);
-      }
-    }
+    functor::Roll<Device, T>()(context, num_elements, num_dims, dim_size,
+            input_flat, output_flat, threshold, dim_range, isd);
   }
 };
 
+namespace functor {
+template<typename T>
+struct Roll<CPUDevice, T> {
+  void operator()(const OpKernelContext *context,
+                  const int64 num_elements,
+                  const int num_dims,
+                  const gtl::ArraySlice<int32> dim_size,
+                  const T *input, T *output,
+                  const gtl::ArraySlice<int32> threshold,
+                  const gtl::ArraySlice<int64> dim_range,
+                  const int64 isd) {
+    if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
```

Comment:
Would it be possible to keep the CPU implementations in the functions at their original location? I assume there is no functional change here, but the move makes it hard to tell.

Reply:
@chsigg, thank you for your feedback. I just back from holiday and will work on it in the next few days.

Reply:
@chsigg, I updated the roll_op.cc as you suggested.